articles/tcp-node/index.md

```diff
@@ -0,0 +1,233 @@
+
+```JavaScript
+// server.js
+const net = require("net"); // import net
+
+// create the server
+let server = net.createServer(connection => {
+    // run all of this when a client connects
+    console.log("new connection");
+});
+```
+
+Great! Let's try running it and... nothing happens. We didn't ask the server to actually wait for a connection. To do this, we'll use `server.listen`. We'll plug in 50,000 as the port number.
+
+```JavaScript
+// server.js
+
+// ...
+
+// look for a connection on port 50,000
+server.listen(50000, () => {
+    console.log("waiting for a connection"); // prints on start
+});
+```
+
+Now that the server can connect to a client, we want it to do something once it receives data. The net library uses an event system. Inside the block where we define the server, we want to make an event when data is received.
+
+```JavaScript
+// server.js
+const net = require("net"); // import net
+
+// create the server
+let server = net.createServer(connection => {
+    // run all of this when a client connects
+    console.log("new connection");
+
+    connection.on("data", data => {
+        // run this when data is received
+        console.log(data.toString()); // prints the data
+    });
+});
+
+// ...
+```
+
+We want to do a little more than print out the data though. We want to an actual calculation. Now, we should implement the protocol in our server.
+
+```JavaScript
+//...
+    connection.on("data", data => {
+        // run this when data is received
+        const dataArgs = data.split(" "); // splits the data into spaces
+
+        if (dataArgs.length === 0) { // in case there is no command
+            connection.write("ERROR no data");
+            return; // prevents other code from running
+        }
+
+        const command = dataArgs[0]; // gets the command
+
+        if (command === "ADD") { // add command
+            if (dataArgs !== 3) { // in case there aren't enough arguments
+                connection.write("ERROR incorrect number of arguments");
+                return;
+            }
+
+            const op1 = Number.parse(dataArgs[1]); // first number
+            const op2 = Number.parse(dataArgs[2]); // second number
+            const result = (op1 + op2).toString(); // result as a string
+
+            connection.write("RESULT " + result);
+            break; // end
+        } else { // invalid command
+            connection.write("ERROR invalid command");
+            return;
+        }
+    });
+
+// ...
+```
+
+You can run this by typing `node server.js` in a command prompt or a terminal.
+
+### Client
+Now that the server is done, we need a program to send information. It needs to connect to the server.
+
+```JavaScript
+// client.js
+const net = require("net"); // import net
+const readline = require("readline").createInterface({
+        input: process.stdin,
+        output: process.stdout
+}); // this will be important later
+
+const options = {
+    port: 9090,
```

Comment:
it looks like the server is listening to a client connecting on port 50000, not 9090?